SCAutolib/models/card.py

```diff
@@ -0,0 +1,229 @@
+
+
+class Card:
+    """
+    Interface for child classes. All child classes will rewrite common methods
+    based on the type of the card.
+    """
+    uri: str = None
+    user: User = None
+    _pattern: str = None
+
+    def _set_uri(self):
+        """
+        Sets URI for given smart card. Uri is matched from ``p11tool`` command
+        with regular expression. If URI is not matched, assertion exception
+        would be raised
+
+        :raise: AssertionError
+        """
+        cmd = ["p11tool", "--list-token-urls"]
+        out = run(cmd).stdout
+        urls = re.findall(self._pattern, out)
+        assert len(urls) == 1, "More URLs are matched"
+        self.uri = urls[0]
+
+    def insert(self):
+        """
+        Insert the card.
+        """
+        ...
+
+    def remove(self):
+        """
+        Remove the card.
+        """
+        ...
+
+    def enroll(self):
+        """
+        Enroll the card (upload a certificate and a key on it)
+        """
+        ...
+
+
+class VirtualCard(Card):
+    """
+    This class provides method for manipulating with virtual smart card. Virtual
+    smart card by itself is represented by the systemd service in the system.
+    The card relates to some user, so providing the user is essential for
+    correct functioning of methods for the virtual smart card.
+
+    Card root directory has to be created before calling any method
+    """
+
+    _service_name: str = None
+    _service_location: Path = None
+    _softhsm2_conf: Path = None
+    _nssdb: Path = None
+    _template: Path = Path(TEMPLATES_DIR, "virt_cacard.service")
+    _pattern = r"(pkcs11:model=PKCS%2315%20emulated;" \
+               r"manufacturer=Common%20Access%20Card;serial=.*)"
+
+    def __init__(self, cert: Path, key: Path, user: User, insert: bool = False,
+                 softhsm2_conf: Path = None):
+        """
+        Initialise virtual smart card. Constructor of the base class is also
+        used.
+
+        :param cert: Path to the certificate for this card
+        :type cert: pathlib.Path
+        :param key: Path to private key for this card
+        :type key: pathlib.Path
+        :param user: User of this card
+        :type user: User
+        :param insert: If True, the card would be inserted on entering the
+            context manger. Default False.
+        :type insert: bool
+        :param softhsm2_conf: path to SoftHSM2 configuration file
+        :type softhsm2_conf: pathlib.Path
+        """
+
+        self.user = user
+        assert self.user.card_dir.exists(), "Card root directory doesn't exists"
+
+        self._private_key = key
+        self._cert = cert
+
+        self._service_name = f"virt_cacard_{self.user.username}"
+        self._service_location = Path(
+            f"/etc/systemd/system/{self._service_name}.service")
+        self._insert = insert
+        self._nssdb = self.user.card_dir.joinpath("db")
+        self._softhsm2_conf = softhsm2_conf
```

Comment:
So currently softhsm2_conf can be any path, be we want it to be in the users home directory right? Maybe that is something we can set directly without asking it as argument?

Reply:
well, it makes sense to look for softhsm2 config file in some default path as users home dir, but I would still leave possibility to set other path

Reply:
I tend to agree with Pavel that this method should be able to work with any path and specific path should probably be supplied by controller.

Reply:
I'm alright with that